File: solutions/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/solution.py
class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        self.count = 0

        def _btrack(arr):
            l = len(arr)
            if(l == n):
                self.count += 1
                if(self.count == k):
                    return ''.join(arr)
                else:
                    return ""

            last = ""
            if(l > 0):
                last = arr[-1]
            
            for char in "abc":
                if char != last:
                    arr.append(char)
                    result = _btrack(arr)
                    if(result != ""):
                        return result

                    arr.pop()
            
            return ""

        return _btrack([])

File: solutions/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/solution.js
/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getHappyString = function(n, k) {
    let count = 0;

    return btrack([]);

    // recursively make string
    function btrack(s){
        // if string is target length, increment our count
        if(s.length === n){
            count += 1
            if(count === k){
                return s.join("")
            }else{
                return "";
            }

        }
        const last = s[s.length - 1];

        for(const char of "abc"){
            if(char !== last){
                s.push(char);
                const res = btrack(s);
                if(res !== ""){
                    return res;
                }
                s.pop();
            }
        }

        return "";
    }
};
